Improvements to excess IP release handshake

hemanthmalla opened the pull request:
Follow up from https://github.com/cilium/cilium/pull/17939 

* Fix for blocked state transition from ready-for-release to released
* Fix for unnecessary updates between agent and operator during handshake

hemanthmalla committed:
Improvements to IP release handshake

* Fix for blocked state transition from ready-for-release to released
* Fix for unnecessary updates between agent and operator during handshake

Signed-off-by: Hemanth Malla <hemanth.malla@datadoghq.com>

gandro commented on pkg/ipam/node.go:
```diff
@@ -794,11 +796,14 @@ func (n *Node) MaintainIPPool(ctx context.Context) error {
 func (n *Node) PopulateIPReleaseStatus(node *v2.CiliumNode) {
+	// maintainIPPool() might not have run yet since the last update from agent.
+	// Attempt to remove any stale entries
+	n.removeStaleReleaseIPs()
```
This seems to introduce a data race, (go test -race inside `pkg/ipam`):

I think we need to revisit the locking for the `ipReleaseStatus` map. We seem to be reading it without the mutex, but writing to it with the mutex held, which I don't think is safe.

````
WARNING: DATA RACE
Read at 0x00c00097e3c0 by goroutine 120:
  runtime.mapiterinit()
      /usr/lib/go/src/runtime/map.go:802 +0x0
  github.com/cilium/cilium/pkg/ipam.(*Node).removeStaleReleaseIPs()
      /home/gandro/Isovalent/go/src/github.com/cilium/cilium/pkg/ipam/node.go:577 +0x8c
  github.com/cilium/cilium/pkg/ipam.(*Node).PopulateIPReleaseStatus()
      /home/gandro/Isovalent/go/src/github.com/cilium/cilium/pkg/ipam/node.go:801 +0x52
  github.com/cilium/cilium/pkg/ipam.(*IPAMSuite).TestNodeManagerReleaseAddress.func5()
      /home/gandro/Isovalent/go/src/github.com/cilium/cilium/pkg/ipam/node_manager_test.go:446 +0xaf

Previous write at 0x00c00097e3c0 by goroutine 60:
  runtime.mapassign_faststr()
      /usr/lib/go/src/runtime/map_faststr.go:202 +0x0
  github.com/cilium/cilium/pkg/ipam.(*Node).maintainIPPool()
      /home/gandro/Isovalent/go/src/github.com/cilium/cilium/pkg/ipam/node.go:679 +0x2292
  github.com/cilium/cilium/pkg/ipam.(*Node).MaintainIPPool()
      /home/gandro/Isovalent/go/src/github.com/cilium/cilium/pkg/ipam/node.go:784 +0x130
  github.com/cilium/cilium/pkg/ipam.(*NodeManager).Update.func2()
      /home/gandro/Isovalent/go/src/github.com/cilium/cilium/pkg/ipam/node_manager.go:273 +0x7d
  github.com/cilium/cilium/pkg/trigger.(*Trigger).waiter()
      /home/gandro/Isovalent/go/src/github.com/cilium/cilium/pkg/trigger/trigger.go:198 +0x35b
  github.com/cilium/cilium/pkg/trigger.NewTrigger·dwrap·1()
      /home/gandro/Isovalent/go/src/github.com/cilium/cilium/pkg/trigger/trigger.go:119 +0x39
```

hemanthmalla replied:
@gandro Thanks for catching it, pushed an update to `removeStaleReleaseIPs()` func.

gandro reviewed (changes requested):
Thanks a ton! Unfortunately this approach seems to introduce a data race.

gandro commented on pkg/ipam/node.go:
```diff
@@ -574,12 +574,14 @@ func (n *Node) determineMaintenanceAction() (*maintenanceAction, error) {
 // remove entries from IP release status map in ciliumnode CRD's status. These IPs need to be purged from
 // n.ipReleaseStatus
 func (n *Node) removeStaleReleaseIPs() {
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
```
While this fixes the data race reported in the unit tests, I believe it's still possible to have a data race (apparently not covered by the unit tests) with the `ipReleaseStatus` map. There are two actions which I believe are executed concurrently:

1.  The `ipam-pool-maintainer-nodename` trigger happens whenever the CiliumNode CRD has been updated (trigger happens in `UpdatedResource`). This causes `n.maintainIPPool` to be executed in a Go rotuine (via `n.poolMaintainer -> n.MaintainIPPool -> n.maintainIPPool`). `n.maintainIPPool` performs multiple unprotected operations on `ipReleaseStatus`
https://github.com/cilium/cilium/blob/0c8953f8689ce9c05742197b97c50f802059b121/pkg/ipam/node.go#L613-L615
2. The `ipam-node-k8s-sync-nodename` trigger is triggered every minute via `ipam-node-interval-refresh`. It accesses `ipReleaseStatus`  via `n.k8sSync` -> `n.syncToAPIServer` -> `n.PopulateIPReleaseStatus`:
https://github.com/cilium/cilium/blob/0c8953f8689ce9c05742197b97c50f802059b121/pkg/ipam/node.go#L803

I'm not sure under what conditions `n.ipReleaseStatus` might be nil. Might be worth double checking if that code is still really needed.

hemanthmalla replied:
Thanks for the detailed review. This section in `node.go` is actually not needed anymore. It's now removed.

hemanthmalla committed:
Fix for potential deadlock in TestMarkForReleaseNoAllocate()

crdAllocator.Allocate() aquires lock on allocator first and then on
nodestore. But updateLocalNodeResource() acquires locks in the opposite
order. This commit releases nodestore lock before acquiring allocator lock
to avoid potential deadlocks due to inconsistent lock ordering.

Signed-off-by: Hemanth Malla <hemanth.malla@datadoghq.com>

gandro reviewed (approved):
Awesome! This looks much cleaner than before. Thanks a ton!
